[lukkesreysandeur] iP

lukkesreysandeur opened the pull request:
# Interrobang
> _Simplicity is the ultimate sophistication._ - Leonardo Da Vinci

Do you want to upgrade your efficiency in life? Are you looking for a personal chat bot that is:
* text-based
* ~~Simple~~ SUPER SIMPLE to use
*  AND FREE?!

Well what're you waiting for? You can start using it in 3 simple steps!
1. Download the program from [here](https://github.com/lukkesreysandeur/ip)
2. Open the chat bot.
3. Watch as all your friends gaze upon you in envy :skull::skull:

Features:
- [x] Saving tasks 
- [x] Searching by keyword
- [ ] Search by date
- [ ] Reminders


Or if you want to run the chat bot through the command-line, use `java Duke` in the CLI

---
```java
public static void main() {
    System.out.println("I don't know what else to add so here.");
}
```

Commit:
Add Gradle support

Commit:
Completed Level-1. Greet, Echo, Exit

Commit:
Completed Level-2. Add, List

Commit:
Completed Level-3. Mark as Done.
Added Task class.

Commit:
Completed Level-4.

Commit:
Completed Level-5. Handle errors.
Added DukeExceptions package and classes.
Abstracted functionality from Duke class to lower classes.
Refactored certain components to work with exceptions.

Commit:
Completed Level-6. Delete.
Added new DukeExceptions class.
Fixed regex bug in the ToDoList changeState method.

Commit:
Updated testcase

Commit:
Complete Level-7. Save.
Added StateSaver class.
Added functionality in the Task classes to work with StateSaver.

Commit:
Add Level-8. Dates and Times.

Commit:
Merge branch 'branch-Level-7' into branch-Level-8

Commit:
Merge commit '8476783d3ef99e96eae6246ecfa7270bc77c1975'

Commit:
Merge branch 'branch-Level-8'

Commit:
Merge commit '556af3f47a96b32898ab4cdbd65b16486a4871e8'

Commit:
Add packages, gradle, and more OOP.
Refactored main method logic into Ui, Commands, Parser classes.
Added Commands package.
Added Tasks package.
Renamed StateSaver to Storage.
Renamed ToDoList to TaskList.
Added all classes to packages.

Commit:
Add unit tests for Parser and Duke classes.
Added junit tests for the two classes.
Fixed package names to follow naming convention.

Commit:
Add JavaDocs for some classes.
Added JavaDocs for Duke, Storage, Parser, Ui, TaskList and DateHandler Classes.

Commit:
Add A-JavaDocs.
Added JavaDocs for the rest of the classes, bar the exceptions.

Commit:
Add A-CodingStandard.

Commit:
Add Level-9. Find.
Created FindCommand class.
Added corresponding methods to TaskList class.

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/DateHandler.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/tasks/Deadline.java
#	src/main/java/duke/tasks/Task.java

Commit:
Fix storage bugs.
Fixed bug where state was not saved after executing certain commands.
Added JavaDocs for previously added methods.

euph00 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    public void printResponse(String input) {
+        String horizLine = "_____________________________";
+        System.out.printf("%s\n%s\n%s\n%n", horizLine, input, horizLine);
```
good idea for formatting strings, I'll be borrowing this formatting

securespider commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,166 @@
+package duke;
+
+import duke.exceptions.DukeEmptyInputException;
```
Good that all Exceptions are explicitly mentioned

securespider commented on src/main/java/duke/DateHandler.java:
```diff
@@ -0,0 +1,30 @@
+ * @author lukkesreysandeur
+ */
+public class DateHandler {
+    private static DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyyMMdd");
```
Instead of having a fixed pattern inside date handler, maybe it might be preferable to have multiple types of patterns and store them in an enum class in case the user were to put in different date patterns

lukkesreysandeur replied:
Ah thats a good suggestion. I will try implementing that, thanks!

securespider commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,52 @@
+
+import duke.commands.Command;
+import duke.exceptions.DukeUnknownInputException;
+import duke.commands.*;
```
Remember not to use wildcard imports. Explicitly say each package you import.
IntelliJ might have auto merged them, you can change this in the settings

lukkesreysandeur replied:
I see I think IntelliJ must have auto merged them. Thanks for letting me know.

securespider commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,97 @@
+        String data = lst.listToText();
+        try {
+            FileWriter writer = new FileWriter(saveLocation);
+            writer.write(data);
```
The error might have occurred during the write statement, which cause the FileWriter object to not be closed properly. Might be better to put in separate try catch statements or use "finally" to close the FileWriter

securespider reviewed (commented):
Looks good!
Just a few comments for your considerations

Commit:
Add Level-10. GUI
Added basic GUI for the Duke chatbot.

lukkesreysandeur committed:
Merge pull request #1 from lukkesreysandeur/branch-Level-10

Merge Level-10. GUI

Commit:
Update Storage to use assertions.
Storage methods use exceptions to handle assumptions - files existing.

The use of exceptions makes it seem like the file not existing is a
condition that could potentially occur.

Extracting this code into assertions makes the code more intuitive as
this behaviour is assumed to be true.

Let's add assertions in the code in places where it is assumed that the
file exists, as it should have been created if it did not.

Commit:
Update all Commands to state assertions.
Commands are assumed to be initialised with non-null inputs.

There is no explicit catch for this, and it is possible that one may
mistakenly use null in their declaration of a command.

Let's add an assertion that the input is not null when initialising the
command.

It is more intuitively stated so that the class is not misused, and if
it is used wrongly, java can catch that mistake.

Commit:
Add A-Assertions.
Cleaned up some commented code.
Refactored Duke's getResponse to properly show the exceptions
that occured.
Added min height to DialogBox to deal with output/input longer than 4
lines.

lukkesreysandeur committed:
Merge pull request #3 from lukkesreysandeur/branch-A-Assertions

Merge branch-A-Assertions

Commit:
Clean up Command code.
Standardise JavaDocs.
Remove unused commented code.

Commit:
Improve comments of Task classes.
Standardise JavaDocs.

Commit:
Update build.gradle.
Update the run configuration in the build.gradle file.

Commit:
Refactor Ui showError method.
showError prints directly to sys.out.

GUI takes input directly, does not print to sys.out.

Let's refactor showError to return the message instead of printing.

Using this method is preferable as the showError method can then be used
in more areas where the GUI is used.

Commit:
Add A-CodeQuality.
Clean up dead code.
Standardise JavaDocs.

lukkesreysandeur committed:
Merge pull request #4 from lukkesreysandeur/branch-A-CodeQuality

Merge branch-A-CodeQuality

Commit:
Update Parser to be more user-friendly.
Parser is not lenient when it comes to capitalisation and whitespace.

This makes the commands very rigid and not friendly to users; users may
be frustrated if their command is refused due to whitespace or caps.

Let's update Parser to ignore trailing whitespace and capitalisation.

Making the Parser more lenient helps to improve the UX.

Commit:
Revert "Update Parser to be more user-friendly."

This reverts commit 43fc9de04635419aacb6f2eb64056bf37280845e.

Commit:
Revert "Revert "Update Parser to be more user-friendly.""

This reverts commit 2ef620463638287412ceef2f7ce49ff5140aca7a.

lukkesreysandeur committed:
Merge pull request #5 from lukkesreysandeur/branch-BCD-Extension

Merge branch-BCD-Extension.

Commit:
Add C-Help.
Add help command to help users with commands.
Improve UX.
Remove dead code.

lukkesreysandeur committed:
Merge pull request #6 from lukkesreysandeur/branch-C-Help

Add C-Help.

Commit:
Improve GUI.
Add A-BetterGui.
Add A-Personality.

Commit:
Refactor classes into packages.
Add all classes to a package except for launcher and Duke.

lukkesreysandeur committed:
Merge pull request #7 from lukkesreysandeur/branch-C-Help

Improve GUI.

Commit:
Add Ui.png

Commit:
Update User Guide.

Commit:
Update User Guide.
Change formatting of the commands section.
